data/data_loader.py

```diff
@@ -38,43 +39,40 @@ def __init__(self, image_train_items: list[ImageTrainItem], seed=555, batch_size
         self.prepared_train_data = image_train_items
         random.Random(self.seed).shuffle(self.prepared_train_data)
         self.prepared_train_data = sorted(self.prepared_train_data, key=lambda img: img.caption.rating())
+        expected_epoch_size = math.floor(sum([i.multiplier for i in self.prepared_train_data]))
+        if expected_epoch_size != len(self.prepared_train_data):
+            logging.info(f" * DLMA initialized with {len(image_train_items)} source images. After applying multipliers, each epoch will train on at least {expected_epoch_size} images.")
+        else:
+            logging.info(f" * DLMA initialized with {len(image_train_items)} images.")
+
         self.rating_overall_sum: float = 0.0
         self.ratings_summed: list[float] = []
         self.__update_rating_sums()
 
-    def __pick_multiplied_set(self, randomizer):
+
+    def __pick_multiplied_set(self, randomizer: random.Random):
         """
         Deals with multiply.txt whole and fractional numbers
         """
-        #print(f"Picking multiplied set from {len(self.prepared_train_data)}")
-        data_copy = copy.deepcopy(self.prepared_train_data) # deep copy to avoid modifying original multiplier property
-        epoch_size = len(self.prepared_train_data)
         picked_images = []
-
-        # add by whole number part first and decrement multiplier in copy
-        for iti in data_copy:
-            #print(f"check for whole number {iti.multiplier}: {iti.pathname}, remaining {iti.multiplier}")
-            while iti.multiplier >= 1.0:
+        fractional_images_per_directory = defaultdict(list[ImageTrainItem])
+        for iti in self.prepared_train_data:
+            multiplier = iti.multiplier
+            while multiplier >= 1:
                 picked_images.append(iti)
-                #print(f"Adding {iti.multiplier}: {iti.pathname}, remaining {iti.multiplier}, , datalen: {len(picked_images)}")
-                iti.multiplier -= 1.0
-
-        remaining = epoch_size - len(picked_images)
-
-        assert remaining >= 0, "Something went wrong with the multiplier calculation"
-
-        # add by remaining fractional numbers by random chance
-        while remaining > 0:
-            for iti in data_copy:
-                if randomizer.uniform(0.0, 1.0) < iti.multiplier:
-                    #print(f"Adding {iti.multiplier}: {iti.pathname}, remaining {remaining}, datalen: {len(data_copy)}")
-                    picked_images.append(iti)
-                    remaining -= 1
-                    iti.multiplier = 0.0
-                if remaining <= 0:
-                    break
-        
-        del data_copy
+                multiplier -= 1
+            # fractional remainders must be dealt with separately
+            if multiplier > 0:
+                directory = os.path.dirname(iti.pathname)
+                fractional_images_per_directory[directory].append(iti)
```

Comment:
This is going to cause problems in the future if (when) multiplier is detached from directory.  I purposely made it a property attached to each image.  The multiplier coming from the directory is an artifact that shouldn't be promulgated throughout the codebase.

Reply:
I don't think the shuffle/bifurcation is the best algorithm here.  Iteration is fast outside some corner cases like someone using a multiplier like 2.00000001 maybe.